File: platform/challenges/svc5-policyforge/service/lib/policy_forge/token.ex
defmodule PolicyForge.Token do
  def sign(secret, user, groups) do
    body = Jason.encode!(%{user: user, groups: groups, exp: System.system_time(:second) + 86_400})
    body64 = Base.url_encode64(body, padding: false)
    sig = :crypto.mac(:hmac, :sha256, secret, body64) |> Base.url_encode64(padding: false)
    body64 <> "." <> sig
  end

  def verify(secret, token) do
    with [body64, sig] <- String.split(token || "", ".", parts: 2),
         expected <- :crypto.mac(:hmac, :sha256, secret, body64) |> Base.url_encode64(padding: false),
         true <- Plug.Crypto.secure_compare(expected, sig),
         {:ok, body} <- Base.url_decode64(body64, padding: false),
         {:ok, json} <- Jason.decode(body),
         true <- json["exp"] >= System.system_time(:second) do
      {:ok, json}
    else
      _ -> :error
    end
  end
end

File: platform/challenges/svc4-nanofleet/service/internal/token/token.go
package token

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"strings"
)

type RouteToken struct {
	Prefix string `json:"prefix"`
}

func Sign(secret, prefix string) string {
	body, _ := json.Marshal(RouteToken{Prefix: prefix})
	body64 := base64.RawURLEncoding.EncodeToString(body)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body64))
	return body64 + "." + base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func Verify(secret, raw string) (RouteToken, bool) {
	parts := strings.Split(raw, ".")
	if len(parts) != 2 {
		return RouteToken{}, false
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(parts[0]))
	want := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	if !hmac.Equal([]byte(want), []byte(parts[1])) {
		return RouteToken{}, false
	}
	body, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return RouteToken{}, false
	}
	var tok RouteToken
	return tok, json.Unmarshal(body, &tok) == nil
}
